Revise premultiply

jcupitt opened the pull request:
This PR revises unpremultiply (again), and tries to simplify thumbnail a bit. It fixes resizing RGBA images in scRGB space, and hopefully doesn't regress an earlier RGBA resize issue. There's a test as well, to try to stop this happening again.

Background: https://github.com/libvips/libvips/issues/1941

Related issue: https://github.com/libvips/libvips/pull/1675

jcupitt committed:
start tinkering

jcupitt committed:
revise unpremultiply, again

We were not detecting division by zero carefully enough, nor clipping
the alpha range sufficiently in unpremultiply.

see https://github.com/libvips/libvips/issues/1941

also see https://github.com/libvips/libvips/pull/1675 for another
difficult test case

jcupitt committed:
add a test for unpremultiply

we actually test vipsthumbnail --linear on an RGBA image, which should
catch everything, hopefully

jcupitt committed:
small comment polish

jcupitt committed:
back to srgb/b-w processing space

for compat in thumbnail behaviour

jcupitt committed:
better again

kleisauke commented on libvips/resample/thumbnail.c:
```diff
@@ -760,16 +785,16 @@ vips_thumbnail_build( VipsObject *object )
+		 * pre-premultiplied format.
 		 */
+		g_info( "premultiplying alpha" ); 
+		have_premultiplied = TRUE;
```
Not sure if this is neater, but instead of keeping a separate `have_premultiplied` flag, you could also set `unpremultiplied_format` to `VIPS_FORMAT_NOTSET` and check for `unpremultiplied_format != VIPS_FORMAT_NOTSET` instead.

jcupitt replied:
Good idea!

kleisauke commented on libvips/resample/thumbnail.c:
```diff
@@ -827,9 +851,10 @@ vips_thumbnail_build( VipsObject *object )
-	else if( thumbnail->export_profile &&
-		(vips_image_get_typeof( in, VIPS_META_ICC_NAME ) || 
-		 thumbnail->import_profile) ) {
+	else if( thumbnail->export_profile ) {
```
Do you think this works when someone does:
```bash
$ vipsthumbnail --export-profile=srgb x.jpg
````
where `x.jpg` is an image _without_ embedded ICC profile? I think the previous check was there to handle this case (perhaps as a speed improvement).

jcupitt replied:
I think this is OK -- `have_imported` is now only set TRUE if `vips_image_get_typeof( in, VIPS_META_ICC_NAME ) || thumbnail->import_profile` earlier in the function, and we only hit this if if `!have_imported`.

jcupitt replied:
But you're right, this won't work if `k2.jpg` has no profile and perhaps it should:

```
$ vipsthumbnail k2.jpg --export-profile cmyk
vipsthumbnail: unable to thumbnail k2.jpg
icc_transform: no input profile
```

It should assume that `k2.jpg` is in srgb.

jcupitt replied:
I committed a thing for this case.

kleisauke commented on libvips/resample/thumbnail.c:
```diff
@@ -844,6 +869,19 @@ vips_thumbnail_build( VipsObject *object )
 	}
+	else if( thumbnail->linear ) {
+		/* Linear mode, no colour management. We went to scRGB for
+		 * precessing, so we now revert to the input
```
nit: typo
```suggestion
		 * processing, so we now revert to the input
```

jcupitt committed:
Update libvips/resample/thumbnail.c

Co-authored-by: Kleis Auke Wolthuizen <github@kleisauke.nl>

jcupitt committed:
cleanup for Kleis comments

jcupitt committed:
Merge branch 'revise-premultiply' of github.com:libvips/libvips into revise-premultiply

jcupitt committed:
allow thumbnail colourspace export with no import

allow this case:

	vipsthumbnail k2.jpg --export-profile cmyk

for an image with no embedded profile

jcupitt committed:
Merge branch '8.10' into revise-premultiply

jcupitt committed:
formatting

jcupitt committed:
Merge branch 'revise-premultiply' of github.com:libvips/libvips into revise-premultiply

jcupitt committed:
Merge branch '8.10' into revise-premultiply

lovell commented:
Profiling suggests the unpremultiply operation might be ~17% slower with this change.

Using callgrind to unpremultiply a 6MP RGBA image with code compiled using gcc 9.3.0 and `-O3`.

Before:
```
  175,523,571  ???:vips_unpremultiply_gen [/usr/local/lib/libvips.so.42.12.7]
```
After:
```
  212,753,893  ???:vips_unpremultiply_gen [/usr/local/lib/libvips.so.42.12.7]
```
Would the following diff be functionally equivalent?
```diff
--- a/libvips/include/vips/util.h
+++ b/libvips/include/vips/util.h
@@ -91,11 +91,7 @@ vips_recip( double x )
 {
        static const double epsilon = 1e-10;
 
-       int sign = x < 0.0 ? -1.0 : 1.0;
-
-       return( sign * x >= epsilon ?
-               1.0 / x :
-               sign / epsilon );
+       return 1 / VIPS_CLIP( -epsilon, x, epsilon );
 }
```
If so, that's only 8% slower:
```
  190,352,189  ???:vips_unpremultiply_gen [/usr/local/lib/libvips.so.42.12.7]
```

jcupitt committed:
remove vips_recip()

it wasn't really necessary, and it was rather slow

jcupitt commented:
I removed the recip thing -- it wasn't really necessary, and it was slower than I expected. Thanks for the speed check!

jcupitt commented:
\o/ good stuff, thanks!
